### PR TITLE
Authentication improvments for 4.0

### PR DESCRIPTION
**Pull request by patrickarlt**
Covers https://github.com/Esri/arcgis-rest-js/issues?q=is%3Aopen+is%3Aissue+milestone%3Av4.0+assignee%3Apatrickarlt Each commit is a separate issue so this can easily reviewed. I plan on handling https://github.com/Esri/arcgis-rest-js/issues/893 in a separate PR.

Summary

* `UserSession` serialize/deserialize now includes the `server` property. #885
* Adds `UserSession.canRefresh` as a getter. #888
* Adds `style` to the oAuth 2.0 options for light/dark themes. #890
* Adds `revokeToken` helper and `UserSession.destroy()`. #892
* Adds a new `UserSession.signIn()` for username/password sign ins. #887
* Adds a new `UserSession.fromToken()` for generating a User Session from an existing token. #887
* `requestOptions.authentication` can now be a plain string. #889
* Rename auth classes for #914:
  * `UserSesison` > `ArcGISIdentityManager`
  * `ApplicationSession` > `ApplicationCredentialsManager`
  * `ApiKey` > `ApiKeyManager`
  Old classes have been deprecated but are still exported. We can remove them at 5.0.

### CODE REVIEW

File: packages/arcgis-rest-request/src/UserSession.ts [hunk 1]
```diff
@@ -697,6 +698,17 @@ export class UserSession implements IAuthenticationManager {
     }
   }
 
+  /**
+   * Revokes all active tokens for a provided {@linkcode UserSession}. The can be considered the equivalent to signing the user out of your application.
+   */
+  public static destroy(session: UserSession) {
```
Comment (tomwayson) [inline packages/arcgis-rest-request/src/UserSession.ts]: a `session.sigOut()` feels better to me than  a static `Session.destroy(session)`, especially as a compliment to `signIn()`. Although, maybe your reasoning was that `Session.signIn()` is just one way to create a session, and `Session.destroy()` works equally well for sessions created w/ `Session.beginOAuth2()`, etc.

This is not something I feel strongly about, and it works for me as is.

File: packages/arcgis-rest-request/src/request.ts [hunk 2]
```diff
@@ -239,6 +240,36 @@ export function request(
...
+    const rawToken = options.authentication;
+
+    authentication = {
+      portal: "https://www.arcgis.com/sharing/rest",
+      getToken: () => {
+        return Promise.resolve(rawToken);
+      }
+    };
+
+    /* istanbul ignore else - we don't need to test NOT warning people */
```
Comment (tomwayson) [inline packages/arcgis-rest-request/src/request.ts:257]: 👍

**Review by tomwayson — APPROVED** [review]
Epic work @patrickarlt 

Thank you for your patience. I'm sorry it took so long to review, but wanted to spend time and go through pretty thoroughly. Despite doing that, I don't have any real improvements to suggest.

The whole `UserSession` -> `ArcGISIdentityManager` is a bit jarring for me. I definitely expect that and other XyzManager classes to be able to juggle more than one set of credentials. That is likely b/c of my long history w/ this library and the ArcGIS API for Javascript.

That said, I don't think my feels on the names should hold anything up (any longer).

LGTM

File: packages/arcgis-rest-request/src/UserSession.ts (hunk 1 above)
Comment (patrickarlt) [reply packages/arcgis-rest-request/src/UserSession.ts]: @tomwayson this is absolutely what I was going for with `destroy` vs `signOut`

Comment (patrickarlt) [reply packages/arcgis-rest-request/src/UserSession.ts]: @tomwayson I would be open to `session.signOut()` as an instance method.

**Pull request merged**